servlet/src/main/java/io/grpc/servlet/ForceTrailersServerInterceptor.java

```diff
@@ -0,0 +1,126 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.servlet;
+
+import io.grpc.ForwardingServerCall;
+import io.grpc.ForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+
+class ForceTrailersServerInterceptor implements ServerInterceptor {
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    ForceTrailersServerCall<ReqT, RespT> interceptedCall = new ForceTrailersServerCall<>(call);
+    try {
+      return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(
+          next.startCall(interceptedCall, headers)) {
+        @Override
+        public void onMessage(ReqT message) {
+          try {
+            super.onMessage(message);
+          } catch (Throwable t) {
+            sendHeaders();
+            throw t;
+          }
+        }
+
+        @Override
+        public void onHalfClose() {
+          try {
+            super.onHalfClose();
+          } catch (Throwable t) {
+            sendHeaders();
+            throw t;
+          }
+        }
+
+        @Override
+        public void onCancel() {
+          try {
+            super.onCancel();
+          } catch (Throwable t) {
+            sendHeaders();
+            throw t;
+          }
+        }
+
+        @Override
+        public void onComplete() {
+          try {
+            super.onComplete();
+          } catch (Throwable t) {
+            sendHeaders();
+            throw t;
+          }
+        }
+
+        @Override
+        public void onReady() {
+          try {
+            super.onReady();
+          } catch (Throwable t) {
+            sendHeaders();
+            throw t;
+          }
+        }
+
+        private void sendHeaders() {
+          interceptedCall.maybeSendEmptyHeaders();
+        }
+      };
+    } catch (RuntimeException e) {
+      interceptedCall.maybeSendEmptyHeaders();
+      throw e;
+    }
+  }
+
+  static class ForceTrailersServerCall<ReqT, RespT> extends
+      ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT> {
+
+    private volatile boolean headersSent = false;
+
+    ForceTrailersServerCall(ServerCall<ReqT, RespT> delegate) {
+      super(delegate);
+    }
+
+    void maybeSendEmptyHeaders() {
+      if (!headersSent) {
+        this.sendHeaders(new Metadata());
+      }
+    }
+
+    @Override
+    public void sendHeaders(Metadata headers) {
+      super.sendHeaders(headers);
+      headersSent = true;
```

Comment:
Move this before the `super.sendHeaders()` "just in case it throws." (Normally I wouldn't care, but we are trying to handle exceptions here, so might as well be consistent.